npms/skale-observer/observer.js

```diff
@@ -232,6 +232,89 @@ async function load_schains( w3, addressFrom, opts ) {
+        if( ! jo_schain )
+            break;
+        arr_schains.push( jo_schain );
+    }
+    return arr_schains;
+}
+
+async function load_schains_connected_only( w3_main_net, w3_s_chain, strChainNameConnectedTo, addressFrom, opts ) {
+    if( ! opts.imaState )
+        throw new Error( "Cannot load S-Chains in observer, no imaState is provided" );
+    if( opts && opts.details )
+        opts.details.write( cc.debug( "Will request all S-Chain(s) hashes..." ) + "\n" );
+    const arrSChainHashes = await opts.imaState.jo_schains_internal.methods.getSchains().call( { from: addressFrom } );
+    const cntSChains = arrSChainHashes.length;
+    if( opts && opts.details )
+        opts.details.write( cc.debug( "Have all " ) + cc.info( cntSChains ) + cc.debug( " S-Chain(s) hashes: " ) + cc.j( arrSChainHashes ) + "\n" );
+    const jo_message_proxy_s_chain =
+        new w3_s_chain.eth.Contract(
+            opts.imaState.joAbiPublishResult_s_chain.message_proxy_chain_abi,
+            opts.imaState.joAbiPublishResult_s_chain.message_proxy_chain_address
+        );
+    const arr_schains = [];
+    for( let idxSChain = 0; idxSChain < cntSChains; ++ idxSChain ) {
+        try {
+            if( opts && opts.bStopNeeded )
+                break;
+            const strSChainHash = arrSChainHashes[idxSChain];
+            const strSChainName = await opts.imaState.jo_schains_internal.methods.getSchainName( strSChainHash ).call( { from: addressFrom } );
+            if( opts && opts.details )
+                opts.details.write( cc.debug( "S-Chain " ) + cc.notice( idxSChain ) + cc.debug( " hash " ) + cc.notice( strSChainHash ) + cc.debug( " corresponds to S-Chain name " ) + cc.notice( strSChainName ) + "\n" );
+            if( opts && opts.bStopNeeded )
+                break;
+            //
+            if( strChainNameConnectedTo == strSChainName ) {
+                if( opts && opts.details )
+                    opts.details.write( cc.debug( "Skip this S-Chain " ) + cc.info( strSChainName ) + cc.debug( " connected status check" ) + "\n" );
+                continue;
+            }
+            if( opts && opts.details ) {
+                opts.details.write(
+                    cc.debug( "Querying connected status between S-Chain " ) + cc.info( strSChainName ) + cc.debug( " and S-Chain " ) +
+                    cc.info( strChainNameConnectedTo ) + cc.debug( "..." ) + "\n" );
+            }
+            const isConnected = await jo_message_proxy_s_chain.methods.isConnectedChain( strSChainName ).call( { from: addressFrom } );
+            if( opts && opts.details )
+                opts.details.write( cc.debug( "Got S-Chain " ) + cc.info( strSChainName ) + cc.debug( " connected status: " ) + cc.yn( isConnected ) + "\n" );
+            if( ! isConnected )
+                continue;
+            const jo_schain = await load_schain( w3_main_net, addressFrom, idxSChain, cntSChains, opts );
```

Comment:
It is possible to pass strSChainHash instead of idxChain to save one eth_call. 
schainsAtSystem(idxChain) would be unnecessary

Reply:
we need idxSChain in call const hash = await opts.imaState.jo_schains_internal.methods.schainsAtSystem( idxSChain ).call( { from: addressFrom } );
but this is discussable)

Reply:
added hash, to make 1 less call, i.e. to skip calling schainsAtSystem